modules/module/LoRAModule.py

```diff
@@ -1,66 +1,76 @@
+import abc
+import enum
+import itertools
 import math
-from abc import ABCMeta
+import typing
 
 import torch
 from torch import nn, Tensor
 from torch.nn import Linear, Conv2d, Parameter
 
+ModuleDict = dict[str, nn.Module]
+StateDict = dict[str, torch.Tensor]
 
-class LoRAModule(metaclass=ABCMeta):
-    prefix: str
-    orig_module: nn.Module
-    lora_down: nn.Module
-    lora_up: nn.Module
-    alpha: torch.Tensor
+class LoRAType(enum.Enum):
+    LierLa = enum.auto()
+    LoCon = enum.auto()
+    LoHa = enum.auto()
+    LoKR = enum.auto()
+    IA3 = enum.auto()
+    DyLoRA = enum.auto()
 
-    def __init__(self, prefix: str, orig_module: nn.Module | None, rank: int, alpha: float):
-        super(LoRAModule, self).__init__()
-        self.prefix = prefix.replace('.', '_')
+class BaseLoRAModule(abc.ABC):
+    type: LoRAType
+    orig_module: nn.Module
+    is_applied: bool
+    prefix: str
+    
+    def __init__(self, prefix: str, orig_module: nn.Module) -> None:
```

Comment:
The orig_module parameter was explicitly defined as `nn.Module | None` instead of `nn.Module`, because in some cases it doesn't exist. When loading only the LoRA wihtout a base model, a DummyLoRAModule is created, which passes None as the orig_module. Same goes for all the lora modules like lora_up, lora_down etc.